File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <algorithm>
using namespace std;
using ll = long long;

const ll MAX = 201010;
ll a[MAX], n, m;

bool decision(ll cur){
    ll cnt = 0, last = -1e12;
    for(int i = 1;i <= n;i++){
        if(a[i] - last < cur) continue;
        last = a[i]; cnt++;
    }

    return cnt >= m;
}

ll maximization(){
    ll lo = 1, hi = 2e9;
    while(lo < hi){
        ll mid = (lo + hi + 1) / 2;
        if(decision(mid)) lo = mid;
        else hi = mid - 1;
    }

    return lo;
}

int main(){
    ios::sync_with_stdio(0); // fastio
    cin.tie(0), cout.tie(0); // fastio

    cin >> n >> m;
    for(int i = 1;i <= n;i++) cin >> a[i];
    sort(a + 1, a + n + 1);

    cout << maximization();

    return 0;
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

n, m = list(map(int, input().rstrip().split()))
a = []
for _ in range(n):
    x = int(input().rstrip())
    a.append(x)
a.sort() 

def decision(cur):
    cnt = 0
    last = -1e12

    for i in a:
        if i - last < cur:
            continue
        last = i
        cnt += 1

    return cnt >= m

def maximization():
    lo = 1
    hi = int(2e9)
    while lo < hi:
        mid = (lo + hi + 1) // 2
        if decision(mid):
            lo = mid
        else:
            hi = mid - 1
    return lo

print(maximization())